mmpose/models/detectors/voxelpose.py

```diff
@@ -0,0 +1,475 @@
+            space_size (list): Size of the 3D space
+            cube_size (list): Size of the input volume to the  center net.
+            space_center (list): Coordinate of the center of the 3D space
+            sub_space_size (list): Size of the cuboid human proposal.
+            sub_cube_size (list): Size of the input volume to the pose net.
+        project_layer (ConfigDict): Dictionary to construct the project layer.
+        center_net (ConfigDict): Dictionary to construct the center net.
+        center_head (ConfigDict): Dictionary to construct the center head.
+        pose_net (ConfigDict): Dictionary to construct the pose net.
+        pose_head (ConfigDict): Dictionary to construct the pose head.
+        train_cfg (ConfigDict): Config for training. Default: None.
+        test_cfg (ConfigDict): Config for testing. Default: None.
+        pretrained (str): Path to the pretrained 2D model. Default: None.
+        freeze_2d (bool): Whether to freeze the 2D model in training.
+            Default: True.
+    """
+
+    def __init__(self,
+                 detector_2d,
+                 space_3d,
+                 project_layer,
+                 center_net,
+                 center_head,
+                 pose_net,
+                 pose_head,
+                 train_cfg=None,
+                 test_cfg=None,
+                 pretrained=None,
+                 freeze_2d=True):
+        super(VoxelPose, self).__init__()
+
+        self.backbone = builder.build_posenet(detector_2d)
+        if self.training and pretrained is not None:
+            load_checkpoint(self.backbone, pretrained)
+
+        self.freeze_2d = freeze_2d
+        self.project_layer = ProjectLayer(project_layer)
+        self.root_net = CuboidProposalNet(center_net)
+        self.center_head = builder.build_head(center_head)
+        self.pose_net = PoseRegressionNet(pose_net)
+        self.pose_head = builder.build_head(pose_head)
+
+        self.space_size = space_3d['space_size']
+        self.cube_size = space_3d['cube_size']
+        self.space_center = space_3d['space_center']
+
+        self.sub_space_size = space_3d['sub_space_size']
+        self.sub_cube_size = space_3d['sub_cube_size']
+
+        self.num_joints = pose_net['output_channels']
+
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+
+    @staticmethod
+    def _freeze(model):
+        """Freeze parameters."""
+        model.eval()
+        for param in model.parameters():
+            param.requires_grad = False
+
+    def train(self, mode=True):
+        """Sets the module in training mode.
+        Args:
+            mode (bool): whether to set training mode (``True``)
+                or evaluation mode (``False``). Default: ``True``.
+
+        Returns:
+            Module: self
+        """
+        super().train(mode)
+        if mode and self.freeze_2d:
+            self._freeze(self.backbone)
+
+        return self
+
+    def forward(self,
```

Comment:
OpenMMLab codebases have a common top-level model interface design that the `forward_train` and `forward_test` methods should be implemented and invoked in `forward` method. Please check [TopDown (MMPose)](https://github.com/open-mmlab/mmpose/blob/1e94844fb5f118f092ff198a5e51b67d8a71921c/mmpose/models/detectors/top_down.py#L91-L160) or [BaseDetector (MMDetection)](https://github.com/open-mmlab/mmdetection/blob/f08548bfd6d394a82566022709b5ce9e6b0a855e/mmdet/models/detectors/base.py#L156-L174) for reference. Maintaining a consistent interface design is necessary for unified upper-level APIs.

Reply:
Ok